nixos-init: init at 0.1.0 (#433154)
NixOS/nixpkgs: pkgs/by-name/ni/nixos-init/README.md

# nixos-init

A system for the initialization of NixOS.

The most important task of nixos-init is to work around the constraints of the
Filesystem Hierarchy Standard (FHS) that are imposed by other tools (most
importantly systemd itself).

> The primary design principle is to do the minimal work requird to start

Reviewer: ```suggestion
The primary design principle is to do the minimal work required to start
```

systemd. Everything that can be done later SHOULD be done later (i.e. after
systemd has already started). This isn't controversial either, this is the
basic principle behind initrds in the first place.

Extending this init should be done with care and only when strictly necessary.
Extending it should always be a last resort. It is explitly not designed to be

> extended for third party extensions. I would go so far that all third party

Reviewer: ```suggestion
extended by third party extensions. I would go so far that all third party
```

> extensions should be forbidded.

Reviewer: ```suggestion
extensions should be forbidden.
```

> extensions should be forbidded.

Reviewer: This is just a piece of rust code, I wouldn't even know what an "extension", be it third-party or internal would mean. I'd probably rephrase the paragraph, stating adding more functionality to it needs good reason, and it's not meant to be extended dynamically?

> The goal of nixos-init is to eventually entriely replace `activationScripts`

Reviewer: ```suggestion
The goal of nixos-init is to eventually entirely replace `activationScripts`
```

for booting, enable bashless activation and ultimately make NixOS overall more
robust.

## Reasoning

- We already have a native API that can be used to easily extend the system
  (`systemd.services`). This is in all ways superior to the stringified and

>   sequential nature of the `activationScripts`.

Reviewer: Either say "sequential nature of our activation scripts", or call it by its full option name, `system.activationScripts`.

- Most things can be started much later than you would typically assume.
  Because systemd services are parallelized, this should improve startup time.
- Using a fully fledged programming language makes writing correct software

>   easier and should improve the quality of the NixOS boot code.

Reviewer: I'd move this one bullet point up, and maybe add a "For the remaining functionality needed, …" in front.

## Invariants

For now, this does not try to replace all uses of activationScripts. For the
first iteration it only tries to offer an alternative to the prepare-root in
the systemd initrd.

> Never intends to improve or replace scripted initrd. The future of NixOS is
> systemd in the initrd.

Reviewer: ```suggestion
It never intends to improve or replace scripted initrd.
Scripted initrd is being phased out, supporting it is out of scope.
```

## Components

> nixos-init consists of a few components split into separate binaries. Howeer,

Reviewer: ```suggestion
nixos-init consists of a few components split into separate entrypoints. However,
```
